README.adoc

```diff
@@ -47,6 +47,35 @@ endif::[]
 [#general]
 == General Recommendations
 
+
+[#prefer-keyword-arguments]
+=== Prefe Keyword Arguments
+
+If you use positional arguments as the parameters passed to `#perform` it becomes complicated to evolve the interface of the job, i.e. if you add or remove an argument then you need to take extra attention so that the old and new code are compatible, and you can't easily drop an argument in any case, as a job would need to inspect the argument itself to know if it was invoked with the old or new API.
```

Comment:
> becomes complicated to evolve the interface of the job

It's quite tedious to evolve the interface of the job in a zero-downtime compatible way. I can think of an intermediate step with a default kwarg.

> If you use positional arguments ... you need to take extra attention

You need to be extra careful in any case if you change job signature.

The "settle" period can be quite long, with the default Sidekiq retries - 21 days.

Changing job signature deserves its own guideline. I was under the impression we have one, but couldn't quickly find it.

Reply:
yes, but positional arguments make this more complicated compared to kwargs. 

E.g. you with keyword arguments you can do this:
```
1. definition is `perform(a:)` client code does `Job.perform(a: 1)`
2. change definition to`perform(a:, b:)` and client code to `Job.perform(b: 1)`
3. change definition to `perform(b:)`
```
If you get a job scheduled via 1 when the jcode is at 3 you will get a clear error that something was not handled and you can reschedule those. 

Compare with positional arguments

```
1. definition is `perform(a)` client code does `Job.perform(1)`
2. change definition to`perform(a, b)` and client code to `Job.perform(:ignored, 1)`
```
Now you can't change the definition to `perform(b)` because the client code is passing 2 values, so you need a step where you take 2 values but ignore the _second_, and then one more change to remove that one. 

In the meantime you may be handling jobs with the _new_ code which receive as first argument a value scheduled with the _old_ code, and if they happen to be compatible (e.g. both intes) you may disrupt things _silently_.

(Or you can introduce a new argument as "flag", but that in turn makes things more complex, and you might just as well create a new job)

Keyword arguments make this nicer :)

Reply:
>1. definition is `perform(a:)` client code does `Job.perform(a: 1)`
>2. change definition to`perform(a:, b:)`

Wouldn't it blow up with a `ArgumentError: missing keyword: :b` when the previously scheduled job is picked up from the queue?